src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java

```diff
@@ -421,8 +422,54 @@ public Participation findOneByExerciseIdAndStudentLoginAnyState(Long exerciseId,
     }
 
     @Transactional(readOnly = true)
-    public List<Participation> findByCourseId(Long courseId) {
-        return participationRepository.findByCourseId(courseId);
+    public List<Participation> findByCourseIdWithRelevantResults(Long courseId) {
+        List<Participation> participations = participationRepository.findByCourseIdWithEagerResults(courseId);
+        //filter all irrelevant results, i.e. rated = false or before exercise due date
+        for (Participation participation : participations) {
+            List<Result> relevantResults = new ArrayList<Result>();
+            for (Result result : participation.getResults()) {
+                if (result.isRated() == Boolean.FALSE) {
+                    //TODO: for compatibility reasons, we include rated == null, in the future we can remove this
+                    continue;
+                }
+                if (result.getCompletionDate() == null) {
+                    continue;
+                }
+                if (result.getScore() == null) {
+                    continue;
+                }
+                if (participation.getExercise() instanceof QuizExercise) {
```

Comment:
this was a late change and bugfix which I did not expect. I will think how to improve the logic here

Reply:
I actually don't like that we have such a complex logic, but this is due to compatibility reasons, performance reasons and because the exercises are quite different in terms of what kind of results we have. With respect to this, I still think the logic is reasonable. I added more comments so that it might be easier to read